inode.c

```diff
@@ -110,69 +110,118 @@ struct inode *simplefs_iget(struct super_block *sb, unsigned long ino)
+                _bi %= eblock->extents[_ei].ee_len;
+
+            bh2 = sb_bread(sb, eblock->extents[_ei].ee_start + _bi);
             if (!bh2) {
-                brelse(bh);
-                return ERR_PTR(-EIO);
+                ret = -EIO;
+                goto file_search_end;
             }
 
-            dblock = (struct simplefs_dir_block *) bh2->b_data;
 
+            dblock = (struct simplefs_dir_block *) bh2->b_data;
             /* Search file in ei_block */
-            for (fi = 0; fi < dblock->nr_files;) {
-                f = &dblock->files[fi];
-                if (!f->inode) {
+            for (_fi = 0; _fi < dblock->nr_files;) {
+                f = &dblock->files[_fi];
+                if (f->inode && !strncmp(f->filename, dentry->d_name.name,
+                                         SIMPLEFS_FILENAME_LEN)) {
+                    *ei = _ei;
+                    *bi = _bi;
+                    *fi = _fi;
+
+                    brelse(bh);
                     brelse(bh2);
-                    goto search_end;
+                    return 0;
                 }
-                if (!strncmp(f->filename, dentry->d_name.name,
-                             SIMPLEFS_FILENAME_LEN)) {
-                    inode = simplefs_iget(sb, f->inode);
-                    brelse(bh2);
-                    goto search_end;
-                }
-                fi += dblock->files[fi].nr_blk;
+                _fi += dblock->files[_fi].nr_blk;
             }
             brelse(bh2);
-            bh2 = NULL;
         }
     }
+file_search_end:
+    brelse(bh);
+    return ret;
+}
+
+/* Search for a dentry in dir.
+ * Fills dentry with NULL if not found in dir, or with the corresponding inode
+ * if found.
+ * Returns NULL on success, indicating the dentry was successfully filled or
+ * confirmed absent.
+ */
+static struct dentry *simplefs_lookup(struct inode *dir,
+                                      struct dentry *dentry,
+                                      unsigned int flags)
+{
+    struct super_block *sb = dir->i_sb;
+    struct simplefs_inode_info *ci_dir = SIMPLEFS_INODE(dir);
+    struct inode *inode = NULL;
+    struct buffer_head *bh = NULL, *bh2 = NULL;
+    struct simplefs_file_ei_block *eblock = NULL;
+    struct simplefs_dir_block *dblock = NULL;
+    int ei, bi, fi, chk;
+
+    /* Check filename length */
+    if (dentry->d_name.len > SIMPLEFS_FILENAME_LEN)
+        return ERR_PTR(-ENAMETOOLONG);
+
+    /* Read the directory block on disk */
+
+    chk = __file_lookup(dir, dentry, &ei, &bi, &fi);
+    if (chk == -ENOENT)
+        goto search_end; /* Not found, return NULL dentry */
+    if (chk < 0)
+        return ERR_PTR(chk); /* I/O error */
+
+    bh = sb_bread(sb, ci_dir->ei_block);
+    if (!bh)
+        return ERR_PTR(-EIO);
+
+    eblock = (struct simplefs_file_ei_block *) bh->b_data;
+    bh2 = sb_bread(sb, eblock->extents[ei].ee_start + bi);
+    if (!bh2)
+        return ERR_PTR(-EIO);
```

Comment:
Buffer head leak in `simplefs_lookup()`: When `bh2` read fails, `bh` is never released.